enhancements/workload-partitioning/rationalizing-configuration.md

```diff
@@ -0,0 +1,479 @@
+including the machine config with settings for kubelet and CRI-O.
+
+The effect of this decision is that the easiest way to use the
+workload partitioning feature will be through the
+performance-addon-operator. This means that the implementation of the
+feature is split across a few components, which avoids duplicating a
+lot of the work already done in the PAO.
+
+Workload partitioning is currently enabled as part of installing a
+cluster, and that will not change as part of the implementation of
+this design. Enabling partitioning early in this way provides
+predictable behavior from the beginning of the life-cycle, and avoids
+extra reboots required to enable it in a running cluster (an important
+consideration in environments with short, fixed maintenance
+windows). To continue to support enabling the feature during
+deployment, and to simplify the configuration, we propose to extend
+the PAO with a `render` command, similar to the one in other
+operators, to have it generate manifests for the OpenShift installer
+to consume. This avoids any need to change the OpenShift installer to
+support this uncommon use case.
+
+The `render` command will have an option to enable the management
+partition. When the option is given, the `render` command will create
+a manifest for the new enablement API with status data that includes
+the `management` workload type to enable partitioning (the details of
+that API will be defined in a separate design document). In the
+future, the PAO might enable partitioning for all of the types
+mentioned in the input PerformanceProfiles, or based on some other
+input.
+
+For each PerformanceProfile, the PAO will render MachineConfig
+manifests with higher precedence than the default ones created by the
+installer during bootstrapping. These manifests will include all of
+the details for configuring `kubelet`, CRI-O, and `tuned` to know
+about and partition workloads, including the cpusets for each workload
+type.
+
+The PAO `render` command also generates a manifest with the CRD for
+PerformanceProfile so bootstrapping does not block when trying to
+install the PerformanceProfile manifests.
+
+### Future Work
+
+1. Write an enhancement to describe the API the admission hook will
+   use to determine when workload partitioning is enabled.
+2. Write an enhancement to describe the changes in
+   performance-addon-operator to manage the kubelet and CRI-O
+   configuration when workload partitioning is enabled, including the
+   `render` command. See https://issues.redhat.com/browse/CNF-2164 to
+   track that work.
+
+### User Stories
+
+#### Enabling and Configuring at the Same Time
+
+In this workflow, the user provides all of the information needed to
+fully partition the workloads from the very beginning.
+
+1. User runs the installer to `create manifests` to get standard
+   manifests without any partitioning.
+2. The user creates PerformanceProfile manifests for each known
```

Comment:
Wouldn't we need those CR's to be applied though ?

Reply:
It does seem like we would normally want the PerformanceProfiles added to the cluster so that when the PAO is installed it has the same inputs available as the render command. What would be the benefit of not installing them?

I think we discussed it, but I don't remember the answer. Does the PAO need any of the other manifests the installer will create in step 1? If not, should we move step 3 before step 1 to make that clear?

Reply:
YEs, I think even the PerformanceProfile needs passed to the installer. For multiple reasons: all the generated objects will have an ownership reference pointing to it, and the user will not want to post it again post install (even though it would probably work too).

About the order, PAO does not need anything from the installer (yet), it is not reading anything. That might change as we might be checking the MCPs and Nodes to generate the right label selectors.

Reply:
My only concern is the PerformanceProfile CRD. That one is normally owned by OLM...

Reply:
If `render` includes the CRD, then either the OLM version would overwrite it or OLM could simply update the ownership like some other operators do. We should verify that that works, though, in case we need to teach OLM that some things it installs might already exist.